### PR TITLE
Handle transactions towards metachain in DbLookupExtensions

### PR DESCRIPTION
**Pull request by andreibancioiu**
The metachain directly notarizes miniblocks towards it - at commit block time. Therefore, there's the time to set the hyperblock coordinates for the miniblocks in question. At the source shard, the "block notarized" notifications arrive in a slightly different manner - the payload is a `block.Header`, not a `block.MetaBlock`.

Therefore,

 - Gather hyperblock coordinates for transactions towards meta as well
 - Refactoring: separate normal from invalid transactions in `nodeTransactions.go`
 - Return extra fields for `unsigned` transactions (e.g. return message, original transaction hash)
 - Improve transaction status (API) reporting when `DatabaseLookupExtensions` are enabled (that is, return `executed` if hyperblock coordinates are fully set, that is, if parent miniblock is fully notarized)
 - Fix value of transaction type (API) when `DatabaseLookupExtensions`, and miniblock metadata is present.

How to test:
 - Send a staking transaction, then inspect its hyperblock coordinates
 - Send invalid transaction, check its Type field
 - Also inspect hyperblock coordinates for some rewards transactions (as a regression test)

### CODE REVIEW

File: data/transaction/status.go [hunk 1]
```diff
@@ -23,12 +23,13 @@ const (
...
 // StatusComputer computes a transaction status
 type StatusComputer struct {
-	MiniblockType    block.Type
-	SourceShard      uint32
-	DestinationShard uint32
-	Receiver         []byte
-	TransactionData  []byte
-	SelfShard        uint32
+	MiniblockType             block.Type
+	IsMiniblockFullyNotarized bool
```
Comment (sasurobert) [inline data/transaction/status.go:27]: IsMiniblockFinalized ?

**Review by sasurobert — DISMISSED** [review]

File: data/transaction/status.go (hunk 1 above)
Comment (LucianMincu) [reply data/transaction/status.go:27]: @sasurobert Isn't the block finalized only at n+1 ?

Comment (sasurobert) [reply data/transaction/status.go:27]: IsMiniBlockFully notarized is for when even the destination processes the miniblock and the result of the process is finalized, included on meta, and meta is finalized.

Comment (andreibancioiu) [reply data/transaction/status.go]: Fixed.

**Review by LucianMincu — APPROVED** [review]
System tests passed.

File: core/dblookupext/historyRepository.go [hunk 2]
```diff
@@ -228,27 +239,45 @@ func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.He
 	for i, header := range headers {
 		headerHash := headersHashes[i]
 
-		metaBlock, ok := header.(*block.MetaBlock)
-		if !ok {
+		asMetaBlock, isMetaBlock := header.(*block.MetaBlock)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go:242]: Renamed.

File: core/dblookupext/historyRepository.go [hunk 3]
```diff
@@ -228,27 +239,45 @@ func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.He
...
+		if isMetaBlock {
+			for _, shardData := range asMetaBlock.ShardInfo {
+				hr.onNotarizedInMetaBlock(asMetaBlock.GetNonce(), headerHash, &shardData)
+			}
 			return
 		}
 
-		for _, shardData := range metaBlock.ShardInfo {
-			hr.onNotarizedBlock(metaBlock.GetNonce(), headerHash, shardData)
+		asBlock, isBlock := header.(*block.Header)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go:250]: Renamed.

File: core/dblookupext/historyRepository.go [hunk 4]
```diff
@@ -228,27 +239,45 @@ func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.He
...
+		if isBlock {
+			if asBlock.GetShardID() != core.MetachainShardId {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go:252]: Removed then.

File: core/dblookupext/historyRepository.go [hunk 5]
```diff
@@ -228,27 +239,45 @@ func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.He
...
 		headerHash := headersHashes[i]
 
-		metaBlock, ok := header.(*block.MetaBlock)
-		if !ok {
+		asMetaBlock, isMetaBlock := header.(*block.MetaBlock)
+		if isMetaBlock {
+			for _, shardData := range asMetaBlock.ShardInfo {
+				hr.onNotarizedInMetaBlock(asMetaBlock.GetNonce(), headerHash, &shardData)
+			}
 			return
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go:247]: @SebastianMarian, not sure what you mean - we iterate on the whole list of `ShardInfo`s. What is not needed, it's discarded in `onNotarizedMiniblock`, of course - see logic around `iDontCare`.

File: core/dblookupext/historyRepository.go [hunk 6]
```diff
@@ -228,27 +239,45 @@ func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.He
...
 
-		for _, shardData := range metaBlock.ShardInfo {
-			hr.onNotarizedBlock(metaBlock.GetNonce(), headerHash, shardData)
+		asBlock, isBlock := header.(*block.Header)
+		if isBlock {
+			if asBlock.GetShardID() != core.MetachainShardId {
+				continue
+			}
+
+			hr.onNotarizedInRegularBlockOfMeta(asMetaBlock.GetNonce(), headerHash, asBlock)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go:256]: Good catch :partying_face: 

My unit test was hiding that - see what I've changed there :exploding_head:

File: core/dblookupext/historyRepository.go [hunk 7]
```diff
@@ -228,27 +239,45 @@ func (hr *historyRepository) onNotarizedBlocks(shardID uint32, headers []data.He
...
+			return
 		}
+
+		log.Error("onNotarizedBlocks(): unexpected type of header", "type", fmt.Sprintf("%T", header))
 	}
 }
 
-func (hr *historyRepository) onNotarizedBlock(metaBlockNonce uint64, metaBlockHash []byte, blockHeader block.ShardData) {
+func (hr *historyRepository) onNotarizedInMetaBlock(metaBlockNonce uint64, metaBlockHash []byte, blockHeader *block.ShardData) {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go:264]: Renamed.

**Review by sasurobert — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed.

File: node/nodeTransactions.go [hunk 8]
```diff
@@ -218,11 +226,13 @@ func (n *Node) getTxBytesFromStorageByEpoch(hash []byte, epoch uint32) ([]byte,
...
 	switch txType {
 	case transaction.TxTypeNormal:
-		fallthrough
-	case transaction.TxTypeInvalid:
 		if tx, ok := txObj.(*transaction.Transaction); ok {
 			return n.prepareNormalTx(tx)
 		}
+	case transaction.TxTypeInvalid:
+		if tx, ok := txObj.(*transaction.Transaction); ok {
+			return n.prepareInvalidTx(tx)
```
Comment (iulianpascalau) [inline node/nodeTransactions.go:234]: :+1:

**Review by iulianpascalau — APPROVED** [review]

File: core/dblookupext/historyRepository.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go:247]: Oh, now I see. Will fix.

[the remark this answers is not in the document]
Comment (andreibancioiu) [reply core/dblookupext/historyRepository.go]: Fixed.

**Review by SebastianMarian — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]

**Pull request merged**